### ISSUE TITLE
No documentation on dotnet/mono framework.

### ISSUE THREAD

**Issue by wizzup** [NixOS/nixpkgs#28294]
## Issue description
I am learning about using dotnet/mono on nixos but can't find any information in [documentation](https://nixos.org/nixpkgs/manual/)

Related files:
* pkgs/top-level/dotnet-packages.nix
* pkgs/build-support/build-dotnet-package/default.nix
* nixpkgs/pkgs/build-support/dotnetenv/default.nix

**Comment by disassembler** [comment]
It would be great to have some documentation, although I'm guessing there aren't a whole lot of people using dotnet on nixos. Here's a slide deck I found that has some info that might help you out though.

https://www.slideshare.net/sandervanderburg/vanderburg10-dotnetnix

**Comment by baracoder** [comment]
Stumbled on this issue by accident (triage). By now we have in addition to mono a dotnet-sdk package.
I am interested in building and running dotnet-sdk on Linux, which seams to be a different use case then described in the mentioned slides. I have started working on a tool which could help building dotnet-sdk projects with NuGet dependencies. https://github.com/baracoder/dotnet2nix

**Comment by teto** [comment]
I was interested in packaging the microsfot python LSP server but I hit the absence of doc too.

**Comment by r-k-b** [comment]
still important to me; the kind of documentation I'm looking for would cover what to do with errors such as `MSB4242: The SDK resolver "Microsoft.DotNet.MSBuildSdkResolver" failed to run` and `You must add a reference to assembly netstandard, Version=2.0.0.0`

**Comment by jonringer** [comment]
I added some initial documentation here: https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/dotnet.section.md

**Comment by jonringer** [comment]
contributions are welcome.

**Comment by YoshiRulz** [comment]
I'm interested in packaging libraries, but there's little documentation or pre-existing code to reference, and .NET Framework (Mono runtime) isn't "first-class" alongside .NET Core. (Is there a better term for that?) I think .NET language support could use some direction at a high level.

**Comment by jonringer** [comment]
> I'm interested in packaging libraries, but there's little documentation or pre-existing code to reference, and .NET Framework (Mono runtime) isn't "first-class" alongside .NET Core. (Is there a better term for that?) I think .NET language support could use some direction at a high level.

Agreed, but in nixpkgs, it largely just takes one person with some domain expertise and will to implement something.

**Comment by kirinnee** [comment]
> Here is some documentation for the new `buildDotnetModule` framework, in case that helps: https://nixos.org/manual/nixpkgs/unstable/#packaging-a-dotnet-application.
> 
> It's not very thorough just yet, but I'll add to it in the future.

How does this work with fetchNuget functionality?
Are there easy ways to convert already publish Nuget packages (such as global tools)?  

Apologise ahead of time, I am not too familiar with Nix itself

**Comment by IvarWithoutBones** [comment]
> How does this work with fetchNuget functionality?

It doesn't use `fetchNuget` right now, instead we create our own nix-based lockfile containing version information and hashes for nuget dependencies.

> Are there easy ways to convert already publish Nuget packages (such as global tools)?

You could set the [`packNupkg`](https://nixos.org/manual/nixpkgs/unstable/#packaging-a-dotnet-application) attribute if you can compile it from source, but installing it as a dotnet tool might take a bit more effort. We currently don't have any setup in place for it, although I do plan to get on that in the future. Would you mind sharing what tool you need, and how you would like to use it? I'd be happy to test it while adding the functionality to install global tools to `buildDotnetModule`. 

> Apologise ahead of time, I am not too familiar with Nix itself

No need to apologise! Always feel free to ask any questions :blush:

**Comment by kirinnee** [comment]
Thank you so much for your reply!

I'd like to use `dotnet-ef` and I've already made a packaging request about it [here](https://github.com/NixOS/nixpkgs/issues/161332). Personally, I have been testing on how to package the simplest CLI tool, [dotnetsay](https://www.nuget.org/packages/dotnetsay/).

I am fairly limited in my knowledge of the inner workings and best-practice of .NET landscape, but would it be possible to use `makeWrapper` and use `${pkgs.dotnet}/bin ${self}/lib/..../tool.dll` where self is the `fetchNuget` already builds all `dll`. I assume this would make it platform and OS independent. What would be the caveats around this method?

**Comment by IvarWithoutBones** [comment]
> I am fairly limited in my knowledge of the inner workings and best-practice of .NET landscape, but would it be possible to use makeWrapper and use ${pkgs.dotnet}/bin ${self}/lib/..../tool.dll where self is the fetchNuget already builds all dll. I assume this would make it platform and OS independent. What would be the caveats around this method?

That would use the cross platform executable, instead of a native one. Not a huge issue, but it means startup times will be a bit slower. It's also not building from source, but instead using a fetched DLL. Should work fine as a temporary workaround though :+1:
